src/tokens/ERC5095.sol

```diff
@@ -248,6 +350,120 @@ contract ERC5095 is ERC20Permit, IERC5095 {
         return returned;
     }
 
+    /// @notice At or after maturity, burns `a` PTs from owner and sends underlying to `receiver`. Before maturity, sends `assets` by selling shares of PT on a YieldSpace AMM.
+    /// @param a The amount of underlying tokens withdrawn
+    /// @param r The receiver of the underlying tokens being withdrawn
+    /// @param o The owner of the underlying tokens
+    /// @param m Minimum amount of shares to be received
+    /// @return uint256 The amount of principal tokens burnt by the withdrawal
+    function withdraw(
+        uint256 a,
+        address r,
+        address o,
+        uint256 m
+    ) external returns (uint256) {
+        // Refers to how many shares are needed to withdraw `a` underlying
+        uint128 shares = Cast.u128(previewWithdraw(a));
+
+        // Confirm that slippage has not been exceeded
+        if (m > shares) {
+            revert Exception(16, m, shares, address(0), address(0));
+        }
+
+        // Pre maturity
+        if (block.timestamp < maturity) {
+            // Receive the shares from the caller
+            _transfer(o, address(this), shares);
+
+            // If owner is the sender, sell PT without allowance check
+            if (o == msg.sender) {
+                uint128 returned = IMarketPlace(marketplace).sellPrincipalToken(
+                    underlying,
+                    maturity,
+                    shares,
+                    Cast.u128(a - (a / 100))
```

Comment:
here I think we use `buyUnderlying` and we use `m` within it to cut the slippage check on 369.

Similarly, it allows us to use a direct pattern from params rather than any calculated values.
```
uint128 returned = IMarketPlace(marketplace).buyUnderlying(
                    underlying,
                    maturity,
                    a,
                    m)
```